src/Confluent.Kafka/ProducerBuilder.cs

```diff
@@ -130,6 +174,38 @@ public ProducerBuilder(IEnumerable<KeyValuePair<string, string>> config)
             return this;
         }
 
+        /// <summary>
+        ///     Set a custom partitioner to use when producing messages to
+        ///     <paramref name="topic" />.
+        /// </summary>
+        public ProducerBuilder<TKey, TValue> SetPartitioner(string topic, PartitionerDelegate partitioner)
+        {
+            if (string.IsNullOrWhiteSpace(topic))
+            {
+                throw new ArgumentException("Topic must not be null or empty");
+            }
+            if (this.Partitioners.ContainsKey(topic))
+            {
+                throw new ArgumentException($"Custom partitioner for {topic} already specified");
+            }
+            this.Partitioners.Add(topic, partitioner);
+            return this;
+        }
+
+        /// <summary>
+        ///     Set a custom partitioner that will be used for all topics
+        ///     except those for which a partitioner has been explicitly configured.
+        /// </summary>
+        public ProducerBuilder<TKey, TValue> SetDefaultPartitioner(PartitionerDelegate partitioner)
```

Comment:
Alternatively, I could have overloaded the name `SetPartitioner`, but i think it adds value to have the word `Default` in there.